### PR TITLE
Implement documentation linter

### PR DESCRIPTION
**Pull request by aslushnikov**
# Documentation Linter

This is a work-in-progress patch that implements documentation linter.

### How it works

The linter works by parsing the `lib/` folder with esprima, parsing `docs/api.md` with markdown-parser, and making sure that classes defined in the `lib/` are getting described in the `docs/api.md`.

The linter describes a bunch of jasmine tests in `lint.js`. 
It could be run via `jasmine test/doclint/lint.js` 

### Challenges
- It turned out to be difficult to find a nice markdown parser that doesn't pull in a whole universe as a dependency. The currently-used [chjj/marked](https://github.com/chjj/marked) is a compromise: it hasn't been updated for a while and doesn't yield token positions in text.

### Room for improvement

- [x] the ESTreeWalker.js is taken from devtools front-end and should be moved under `third_party`
  or substituted with some zero-dependency estree walker. Note though that ESTreeWalker.js walks nodes in the same order as they appear in text
- [ ] method arguments are not getting checked
- [x] we should check for table-of-contents to be up-to-date (and ideally being able to generate it; probably requires a separate script with `--fix` argument)
- [ ] it would be nice to ckeck that methodы have consistent and up-to-date argument description
- [ ] events should be linted as well

References #14.

### CODE REVIEW

File: test/doclint/ESTreeWalker.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2014 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+/**
+ * @unrestricted
```
Comment (JoelEinbinder) [inline test/doclint/ESTreeWalker.js]: Remove @unrestricted.

**Comment by JoelEinbinder** [comment]
Looks good

**Comment by paulirish** [comment]
This isn't run during `yarn test`. Do you want to hook that up? 
Also might as well add a `lint-docs` npm task for it.

**Comment by JoelEinbinder** [comment]
I think that comes after. This patch doesn't pass doclint

**Comment by JoelEinbinder** [comment]
If package-lock.json changes, does yarn.lock also need to change?

**Comment by paulirish** [comment]
yup it should

image

File: test/doclint/ESTreeWalker.js (hunk 1 above)
Comment (aslushnikov) [reply test/doclint/ESTreeWalker.js]: That's actually upstream code, moved it to `third_party`

**Comment by paulirish** [comment]
I've pushed a branch that builds the MDOutline via puppteer and the DOM instead of parsing the marked markdown AST:  https://github.com/GoogleChrome/puppeteer/compare/1a3694c3f1c0ace992b18cba18e4d123017ec9a4...doclint-w-puppeteer

It still has an issue in that `beforeAll` is async and doesn't appear to complete before the following `describe` blocks start executing.

**Comment by aslushnikov** [comment]
Paul, that's a nice idea! This is the way to go, it eliminates so many difficulties with markdown traversing!

I've left a few comments on https://github.com/GoogleChrome/puppeteer/commit/701cdf60276ffec7e407ce520d035e01ee18ed8a and addressed the fixme in https://github.com/GoogleChrome/puppeteer/commit/ef352d634a55db847358828b0e29b82a779df8e7. Could you please rebase your branch atop of master and this PR, and push your changes to the #47?

**Comment by paulirish** [comment]
Branch updated. Lemme try out finishing the TOC.

**Comment by paulirish** [comment]
We now doublecheck the TOC is good:

image


This is ready for review.

File: test/doclint/lint.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  let mdOutline = new MDOutline(apiMdText);
+  await mdOutline.collectHeadings();
+  mdOutline.buildClasses();
+  mdClassesArray = mdOutline.classes;
+}));
+
+describe('table of contents', function() {
+  it('should match markdown-toc\'s output', () => {
+    const newApiMdText = markdownToc.insert(apiMdText);
+    expect(apiMdText === newApiMdText).toBe(true, 'markdown TOC is outdated, run `yarn generate-toc`');
```
Comment (aslushnikov) [inline test/doclint/lint.js:60]: Nicee!

The message could be done slightly more beautiful:
```js
if (apiMdText !== newApiMdText)
   fail('markdown TOC is outdated, run `yarn generate-toc`');
```

File: test/doclint/MDOutline.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+const commonmark = require('commonmark');
+const Browser = require('../../lib/Browser');
+
+class MDOutline {
+  constructor(text) {
+    this.classes = [];
+    this.html = MDOutline.compile(text);
+  }
+
+  static compile(text) {
```
Comment (aslushnikov) [inline test/doclint/MDOutline.js:11]: This could be inlined (or at least looks like a private method)

File: test/doclint/MDOutline.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  async collectHeadings() {
+    const browser = new Browser({args: ['--no-sandbox']});
+    const page = await browser.newPage();
+    await page.setContent(this.html);
+    this.headings = await page.evaluate(getTOCHeadings);
+    await browser.close();
+  }
+
+  buildClasses() {
```
Comment (aslushnikov) [inline test/doclint/MDOutline.js:27]: Let's do a `MDOutline.create(text)` method which will encapsulate buildClasses and collectHeadings methods? The api would be nice:

```js
let mdOutline = await MDOutline.create(text);
```

File: test/doclint/MDOutline.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      if (currentClassName === null)
+        return;
+      this.classes.push(new Documentation.Class(currentClassName, currentClassMethods));
+      currentClassName = null;
+      currentClassMethods = [];
+    }
+  }
+}
+
+function getTOCHeadings(){
```
Comment (aslushnikov) [inline test/doclint/MDOutline.js:71]: I think this function could be slightly simplified:
```js
function getHeadings() {
  let headings = {};
  let methods = [];
  for (let header of body.querySelectorAll('h3,h4')) {
      if (header.matches('h3')) {
          methods = [];
          headings[header.textContent] = methods;
      } else {
          methods.push(header.textContent);
      }
  }
}
```

**Pull request merged**